lib/propolis/src/hw/nvme/queue.rs

```diff
@@ -551,63 +593,55 @@ impl CompQueue {
     fn reserve_entry(
         self: &Arc<Self>,
-        sq: Arc<SubQueue>,
-    ) -> Option<CompQueueEntryPermit> {
-        let mut state = self.state.inner.lock().unwrap();
-        // No more spots available
-        if state.avail == 0 {
-            // Make sure we kick the SQ's when we have space available again
-            state.kick = true;
-
-            None
+        sq: &Arc<SubQueue>,
+    ) -> Option<ProtoPermit> {
+        let mut state = self.state.lock();
+        if state.take_avail() {
+            Some(ProtoPermit::new(self, sq))
         } else {
-            // Otherwise claim a spot
-            state.avail -= 1;
-
-            Some(CompQueueEntryPermit {
-                cq: Arc::downgrade(self),
-                sq: Arc::downgrade(&sq),
-            })
+            // No more spots available.
+            None
         }
     }
 
-    /// Add a new entry to the Completion Queue while consuming a `CompQueueEntryPermit`.
+    /// Add a new entry to the Completion Queue while consuming a `Permit`.
     fn push(
         &self,
-        _permit: CompQueueEntryPermit,
-        entry: RawCompletion,
+        comp: Completion,
+        permit: Permit,
+        sq: &SubQueue,
         mem: &MemCtx,
     ) {
-        // Since we have a permit, there should always be at least
-        // one space in the queue and this unwrap shouldn't fail.
-        let idx = self.state.push_tail().unwrap();
-        probes::nvme_cqe!(|| (self.id, idx, (entry.status_phase & 1) as u8));
-        let addr = self.entry_addr(idx);
-        mem.write(addr, &entry);
-        // XXX: handle a guest addr that becomes unmapped later
-    }
+        let mut cqe = CompletionQueueEntry::new(comp, permit.cid);
+        sq.annotate_completion(&mut cqe);
 
-    /// Returns the current Phase Tag bit.
-    ///
-    /// The current Phase Tag to identify to the host (VM) that a Completion
-    /// entry is new. Flips every time the Tail entry pointer wraps around.
-    ///
-    /// See NVMe 1.0e Section 4.5 Completion Queue Entry - Phase Tag (P)
-    fn phase(&self) -> u16 {
-        let state = self.state.inner.lock().unwrap();
-        state.phase as u16
-    }
+        let mut guard = self.state.lock();
+        let (idx, phase) = guard
+            .push_tail()
+            .expect("CQ should have available space for assigned permit");
+
+        probes::nvme_cqe!(|| (self.id, idx, phase as u8));
 
-    /// Returns the corresponding [`GuestAddr`] for a given entry in
-    /// the Completion Queue.
-    fn entry_addr(&self, idx: u16) -> GuestAddr {
-        let res = self.base.0
-            + idx as u64 * std::mem::size_of::<RawCompletion>() as u64;
-        GuestAddr(res)
+        // The only definite indicator that a CQE has become valid is the phase
+        // bit being toggled.  Since the interface for writing to guest memory
+        // cannot ensure that the other bits of the CQE are written before the
+        // phase bit, we must proceed carefully:
```

Comment:
oof, nice catch!

Reply:
Thanks.

Was it also clear how the old system of querying the phase bit for the CQ outside the lock could result in the wrong phase being set for an entry if two race during completion?  I _think_ I got that right, but there also wasn't a good place to make a comment about it.

Reply:
mmm, that was definitely an oversight in the previous handling.